src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -810,6 +894,162 @@ NodeListInsertCommand(List *workerNodeList)
 }
 
 
+/*
+ * MarkObjectsDistributedCreateCommand generates a command that can be executed to
+ * insert or update the provided objects into pg_dist_object on a worker node.
+ */
+char *
+MarkObjectsDistributedCreateCommand(List *addresses,
+									List *distributionArgumentIndexes,
+									List *colocationIds)
+{
+	StringInfo insertDistributedObjectsCommand = makeStringInfo();
+	int currentObjectCounter;
+
+	Assert(list_length(addresses) == list_length(distributionArgumentIndexes));
+	Assert(list_length(distributionArgumentIndexes) == list_length(colocationIds));
+
+	appendStringInfo(insertDistributedObjectsCommand,
+					 "WITH distributed_object_data(typetext, objnames, "
+					 "objargs, distargumentindex, colocationid)  AS (VALUES ");
+
+	bool isFirstValue = true;
+	for (currentObjectCounter = 0; currentObjectCounter < list_length(addresses);
+		 currentObjectCounter++)
+	{
+		ObjectAddress *address = list_nth(addresses, currentObjectCounter);
+		int32 *distributionArgumentIndex = list_nth(distributionArgumentIndexes,
+													currentObjectCounter);
+		int32 *colocationId = list_nth(colocationIds, currentObjectCounter);
+		List *names;
+		List *args;
+		char *objectType;
+
+		#if PG_VERSION_NUM >= PG_VERSION_14
+		objectType = getObjectTypeDescription(address, false);
+		getObjectIdentityParts(address, &names, &args, false);
+		#else
+		objectType = getObjectTypeDescription(address);
+		getObjectIdentityParts(address, &names, &args);
+		#endif
+
+		if (!isFirstValue)
+		{
+			appendStringInfo(insertDistributedObjectsCommand, ", ");
+		}
+		isFirstValue = false;
+
+		appendStringInfo(insertDistributedObjectsCommand,
+						 "(%s, ARRAY[",
+						 quote_literal_cstr(objectType));
+
+		char *name;
+		bool firstLoop = true;
+		foreach_ptr(name, names)
+		{
+			if (!firstLoop)
+			{
+				appendStringInfo(insertDistributedObjectsCommand, ", ");
+			}
+			firstLoop = false;
+			appendStringInfoString(insertDistributedObjectsCommand, quote_literal_cstr(
+									   name));
+		}
+
+		appendStringInfo(insertDistributedObjectsCommand, "]::text[], ARRAY[");
+
+		char *arg;
+		firstLoop = true;
```

Comment:
minor: lets use more descriptive & seperate variable names as we have multiple loops: `firstInNameLoop`, `firstInArgLoop` and `isFirstDistributedObject` or such